### PR TITLE
Xnor/note names

### PR DESCRIPTION
**Pull request by x37v**
branched off `xnor/const_fn` branch then continued from where you left off, though i had to change the math a little..

added const note values

will move the PR over to main if `xnor/const_fn` gets merged

### CODE REVIEW

File: src/note.rs [hunk 1]
```diff
@@ -2,11 +2,151 @@
...
+    pub const E3: Self = Self::new(4 * 12 + 4);
+    pub const F3: Self = Self::new(4 * 12 + 5);
+    pub const Fs3: Self = Self::new(4 * 12 + 6);
+    pub const G3: Self = Self::new(4 * 12 + 7);
+    pub const Gs3: Self = Self::new(4 * 12 + 8);
+    pub const A3: Self = Self::new(4 * 12 + 9);
+    pub const As3: Self = Self::new(4 * 12 + 10);
+    pub const B3: Self = Self::new(4 * 12 + 11);
+
+    pub const C4: Self = Self::new(5 * 12 + 0);
```
Comment (mendelt) [inline src/note.rs]: I think middle C = C3 = 60 is more common, this is the Yamaha standard where middle C = C4 = 60.
I'm ok with both but maybe we need to document this or something

File: src/note.rs [hunk 2]
```diff
@@ -27,8 +167,132 @@ impl From<u8> for Note {
...
+        (note_name, octave)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum MidiConversionError {
+    OutOfBoundsError,
+}
+
+impl TryFrom<(NoteName, i8)> for Note {
```
Comment (mendelt) [inline src/note.rs]: Maybe we could provide an additional From implementation that just returns the note clamped between min and max. We're doing this everywhere else so maybe provide this here too.

File: src/note.rs (hunk 1 above)
Comment (x37v) [reply src/note.rs]: ah okay, i'm okay with c3 = 60 if it is more common, i'll fix it up

File: src/note.rs (hunk 2 above)
Comment (x37v) [reply src/note.rs]: it looks like rustc won't let you impl both a `TryFrom` and `From` (or `TryInto` and `Into`) to -> from the same type

feels like here, the fallible conversion might be better?

File: src/note.rs (hunk 1 above)
Comment (mendelt) [reply src/note.rs]: Lets check with the usbd guys first, they were using the Yamaha convention too, they might have a good reason for that. (maybe it was standardized in the usb midi spec?)

Also, as this isnt really standardized. Whatever we choose, we should document it somewhere. But I can do that outside of this PR.

File: src/note.rs (hunk 2 above)
Comment (x37v) [reply src/note.rs]: @mendelt i just pushed a change that adds a clamped value to the error and also adds a from to the result type so you can do `.try_into().into()` if you really don't care about the error?

let me know what you think, i can always back it out

Comment (mendelt) [reply src/note.rs]: Can't we just do an `impl From<(NoteName, i8)> for Note` next to the TryFrom? Then you can just choose if you want to do a `.try_into()` or `.into()`. Having to chain them seems to be useless overhead. And it also bloats the Error type if you dont want to use the clamped value.

Comment (x37v) [reply src/note.rs]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f393729cb3505d66f4a97d3ca76b10e

```
conflicting implementation in crate `core`:
          - impl<T, U> TryFrom<U> for T
            where U: Into<T>;
```

https://doc.rust-lang.org/std/convert/trait.TryFrom.html#impl-TryFrom%3CU%3E-for-T

looks like `TryFrom` is automatically implemented if the type implements `Into` which is automatic if you implement `From` so we get a conflict.

Comment (x37v) [reply src/note.rs]: I guess we could do something like this?
```rust
impl TryFrom<(NoteName, i8)> for Note {
    type Error = NoteOutOfBoundsError;

    fn try_from(value: (NoteName, i8)) -> Result<Self, Self::Error> {
        let (name, octave) = value;
        if octave < -2i8 {
            Err(NoteOutOfBoundsError { clamped: Note::MIN })?
        } else if octave > 8 {
            Err(NoteOutOfBoundsError { clamped: Note::MAX })?
        }

        let number = name as u8 + (octave + 2) as u8 * 12;

        if number > 127 {
            Err(NoteOutOfBoundsError { clamped: Note::MAX })
        } else {
            Ok(Note::new(number))
        }
    }
}

impl From<(i8, NoteName)> for Note {
    fn from(value: (i8, NoteName)) -> Note {
        let (octave, name) = value;
        if octave < -2i8 {
            Note::MIN
        } else if octave > 8 {
            Note::MAX
        } else {
            let number = name as u8 + (octave + 2) as u8 * 12;

            if number > 127 {
                Note::MAX
            } else {
                Note::new(number)
            }
        }
    }
}

```

Comment (mendelt) [reply src/note.rs]: We could also just make these extra constructors on Note,
Or even remove the NoteName + octave stuff completely as its not really central to Midi anyway.

Comment (x37v) [reply src/note.rs]: I'm fine with either one of those.. I wouldn't likely use any of the note name stuff myself, but i figure some might find some value in it.

**Comment by x37v** [comment]
I removed the `NoteName` struct but can easily revert to [fac5c8a](https://github.com/rust-midi/midi-types/pull/11/commits/fac5c8a81486bb8aa1d7e90b258d7ceb271ddc39) if that looks better to you?

**Comment by mendelt** [comment]
I like this. Nice and clean. I really like your const solution for the notes. Want to get this merged? Or should we squash some of the commits to get a more readable git-history?

**Comment by x37v** [comment]
> I like this. Nice and clean. I really like your const solution for the notes. Want to get this merged? Or should we squash some of the commits to get a more readable git-history?

I'll squash it to clean it up and merge it myself.
I was thinking I'd like to do a release as well, I want to get a simplified PR over to the `usbd-midi` folks

**Comment by x37v** [comment]
rebased on `main` and cherry picked

**Pull request closed without merging**